tokio-trace/tokio-trace-core/src/lib.rs

```diff
@@ -131,14 +132,16 @@ macro_rules! metadata {
         target: $target:expr,
         level: $level:expr,
         fields: $fields:expr,
-        callsite: $callsite:expr
+        callsite: $callsite:expr,
+        callsite_kind: $callsite_kind:expr,
```

Comment:
The `metadata` macro is included as part of the public API of `tokio-trace-core`. Can we add tests to ensure that the public API has not been broken as part of this change?

Reply:
As I see it, we have a couple options for forwards-compat. For either of these options, we will probably want to add a `CallsiteKind::UNKNOWN` variant or similar.

One option is to change the `metadata!` macro so that instead of requiring a `callsite_kind`, it has two separate match arms, one of which contains a callsite kind, and the other of which does not. The match arm without a callsite kind would expand to invoking the arm with a callsite kind, providing the callsite kind of `UNKNOWN`. Something like this:
```rust
macro_rules! metadata {
    (
        name: $name:expr,
        target: $target:expr,
        level: $level:expr,
        fields: $fields:expr,
        callsite: $callsite:expr,
    ) => {
        metadata! {
            name: $name,
            target: $target,
            level: $level,
            fields: $fields,
            callsite: $callsite,
            callsite_kind: $crate::metadata::CallsiteKind::UNKNOWN,
        }
    };
    // ...
    (
        name: $name:expr,
        target: $target:expr,
        level: $level:expr,
        fields: $fields:expr,
        callsite: $callsite:expr,
        callsite_kind: $callsite_kind:expr,
    ) => {
        // ...
    }
}
```

Alternatively, instead of adding a `callsite_kind` field to `Metadata`, we could add method to get the callsite kind to `Callsite` instead, so that we can have a default impl for backwards-compatibility reasons. That function would default to returning the `UNKNOWN` callsite kind variant, rather than returning incorrect information about old callsites. We could then add methods on `callsite::Identifier` and on `Metadata` so that metadata can expose the callsite's kind.

Personally, I think the first approach is probably better.

Reply:
As an update, we are planning on making a breaking change in the upcoming `tokio-trace-core` release, so I think we may be able to just make a breaking change here, as well. That would mean we wouldn't have to have an `UNKNOWN` callsite kind, which would be nice...

@carllerche do you agree?

Reply:
sounds fine. we should still add API tests 👍